### ISSUE TITLE
gunicorn+nginx queue timeout

### ISSUE THREAD

**Issue by luhn** [benoitc/gunicorn#2260]
I'm using gunicorn+nginx and trying to rig some sort of request queue timeout.  Basically, when the requests start backing up, there comes a point where starting work on a request would be useless, because it's already sat in the queue too long to respond in a timely manner.  I'd like those requests to be dropped, so work can go towards requests that are still fresh.  (Load shedding)

My first thought was nginx's `proxy_connect_timeout`, because AFAIA gunicorn won't accept a connection until a worker is available.  But it appears to have no effect.

My second thought was nginx's `proxy_send_timeout`, because I would think no data could be sent until a worker is ready to accept it.  Still no effect.

I tried setting gunicorn `backlog=0` plus  nginx`proxy_connect_timeout`, which seems to work for most requests, but has some really weird behavior around the edges.  For example I'll send 10 one second requests to gunicorn with one worker, and one of those requests will take 30 seconds and another will read timeout after 60 seconds, even though maximum should be 10s.

If you can't tell, at this point I'm just flailing around with little understanding of what I'm doing.

I was wondering if anybody has experience with this or can think of any possible solutions.

**Comment by luhn** [comment]
An update on where I ended up:

nginx has queuing available in the paid version, unfortunately that's not an option for me.  HAProxy has queuing, but is not able to buffer requests.  So I'm currently running both:  Internet -> nginx for buffering -> HAProxy for load shedding -> gunicorn.

If anybody's interested, I've packaged both my nginx and HAProxy setups as Docker containers.  [gunicorn-proxy](https://github.com/luhn/docker-gunicorn-proxy) does the request buffering and [spillway](https://github.com/luhn/spillway) does the load shedding.
